src/query/functions/temporal/base.go

```diff
@@ -119,191 +128,204 @@ func (c *baseNode) Process(
-) ([]block.UnconsolidatedBlock, bool, error) {
-	numBlocks := bounds.Blocks(leftRangeStart.Start)
-	leftBlks, err := c.cache.multiGet(leftRangeStart, numBlocks, true)
-	if err != nil {
-		return nil, false, err
-	}
-	return leftBlks, len(leftBlks) != numBlocks, nil
-}
+func batchProcess(
+	iterBatches []block.UnconsolidatedSeriesIterBatch,
+	builder block.Builder,
+	m blockMeta,
+	p processor,
+) error {
+	var (
+		metas = m.seriesMeta
 
-// processRight processes blocks after current block. This is done by fetching
-// all contiguous right blocks until the right range.
-func (c *baseNode) processRight(
-	bounds models.Bounds,
-	rightRangeStart models.Bounds,
-) ([]block.UnconsolidatedBlock, bool, error) {
-	numBlocks := rightRangeStart.Blocks(bounds.Start)
-	rightBlks, err := c.cache.multiGet(bounds.Next(1), numBlocks, false)
-	if err != nil {
-		return nil, false, err
-	}
+		mu       sync.Mutex
+		wg       sync.WaitGroup
+		multiErr xerrors.MultiError
+		idx      int
+	)
 
-	return rightBlks, len(rightBlks) != numBlocks, nil
+	builder.PopulateColumns(len(metas))
+	for _, batch := range iterBatches {
+		wg.Add(1)
+		// capture loop variables
+		loopIndex := idx
+		batch := batch
+		idx = idx + batch.Size
+		go func() {
+			err := buildBlockBatch(loopIndex, batch.Iter, builder, m, p, &mu)
+			mu.Lock()
+			// NB: this no-ops if the error is nil.
+			multiErr = multiErr.Add(err)
+			mu.Unlock()
+			wg.Done()
+		}()
+	}
+
+	wg.Wait()
+	return multiErr.FinalError()
 }
 
-func (c *baseNode) propagateNextBlocks(
-	processRequests []processRequest,
-	blocks []block.Block,
-	maxBlocks int,
+func buildBlockBatch(
+	idx int,
+	iter block.UnconsolidatedSeriesIter,
+	builder block.Builder,
+	blockMeta blockMeta,
+	processor processor,
+	mu *sync.Mutex,
 ) error {
-	processedKeys := make([]time.Time, len(processRequests))
+	values := make([]float64, 0, blockMeta.steps)
+	for iter.Next() {
+		var (
+			newVal float64
+			init   = 0
+			end    = blockMeta.end
+			start  = end - blockMeta.aggDuration
+			step   = blockMeta.stepSize
 
-	// propagate blocks downstream
-	for i, nextBlock := range blocks {
-		req := processRequests[i]
-		if err := c.controller.Process(req.queryCtx, nextBlock); err != nil {
-			return err
+			series     = iter.Current()
+			datapoints = series.Datapoints()
+		)
+
+		values = values[:0]
+		for i := 0; i < blockMeta.steps; i++ {
+			iterBounds := iterationBounds{
+				start: start,
+				end:   end,
+			}
+
+			l, r, b := getIndices(datapoints, start, end, init)
+			if !b {
+				newVal = processor.process(ts.Datapoints{}, iterBounds)
+			} else {
+				init = l
+				newVal = processor.process(datapoints[l:r], iterBounds)
+			}
+
+			values = append(values, newVal)
+			start += step
+			end += step
 		}
 
-		processedKeys[i] = req.bounds.Start
+		mu.Lock()
+		// NB: this sets the values internally, so no need to worry about keeping
+		// a reference to underlying `values`.
+		err := builder.SetRow(idx, values, blockMeta.seriesMeta[idx])
```

Comment:
Is the lock required if SetRow(...) only touches that row? (which looking at SetRow it looks like it does?)

I know could be potentially dangerous, although could avoid a lot of locking across batches when processing series after series.

Reply:
Maybe just best to keep the locks..

Reply:
Yeah think it's safer to keep the locks here; applying the actual query is a much heavier process than writing the rows, so any increase is likely negligible